drivers/usb/udc/udc_dwc2.c

```diff
@@ -380,23 +417,66 @@ static inline int dwc2_read_fifo(const struct device *dev, const uint8_t ep,
 	return 0;
 }
 
+static uint32_t dwc2_rx_xfer_size(struct udc_dwc2_data *const priv,
+				  struct udc_ep_config *const cfg,
+				  struct net_buf *buf)
+{
+	uint32_t size;
+
+	if (priv->bufferdma) {
+		size = net_buf_tailroom(buf);
+
+		/* Do as many packets in a single DMA as possible */
+		if (size > priv->max_xfersize) {
+			size = ROUND_DOWN(priv->max_xfersize, cfg->mps);
+		}
+	} else {
+		/* Completer mode can always program Max Packet Size, RxFLvl
+		 * interrupt will drop excessive data if necessary (i.e. buffer
+		 * is too short).
+		 */
+		size = cfg->mps;
+	}
+
+	return size;
+}
+
 /* Can be called from ISR and we call it only when there is a buffer in the queue */
-static void dwc2_prep_rx(const struct device *dev,
+static void dwc2_prep_rx(const struct device *dev, struct net_buf *buf,
 			 struct udc_ep_config *const cfg, const bool ncnak)
 {
 	struct usb_dwc2_reg *const base = dwc2_get_base(dev);
+	struct udc_dwc2_data *const priv = udc_get_private(dev);
 	uint8_t ep_idx = USB_EP_GET_IDX(cfg->addr);
 	mem_addr_t doeptsiz_reg = (mem_addr_t)&base->out_ep[ep_idx].doeptsiz;
 	mem_addr_t doepctl_reg = dwc2_get_dxepctl_reg(dev, ep_idx);
 	uint32_t doeptsiz;
+	uint32_t xfersize;
+
+	xfersize = dwc2_rx_xfer_size(priv, cfg, buf);
 
-	doeptsiz = (1 << USB_DWC2_DOEPTSIZ0_PKTCNT_POS) | cfg->mps;
+	doeptsiz = xfersize | usb_dwc2_set_deptsizn_pktcnt(DIV_ROUND_UP(xfersize, cfg->mps));
 	if (cfg->addr == USB_CONTROL_EP_OUT) {
-		doeptsiz |= (3 << USB_DWC2_DOEPTSIZ0_SUPCNT_POS);
+		/* Use 1 to allow 8 byte long buffers for SETUP data */
+		doeptsiz |= (1 << USB_DWC2_DOEPTSIZ0_SUPCNT_POS);
 	}
 
 	sys_write32(doeptsiz, doeptsiz_reg);
 
+	if (priv->bufferdma) {
+		if (!dwc2_dma_buffer_ok_to_use(dev, buf->data, xfersize, cfg->mps)) {
+			/* Cannot continue unless buffer is bounced. Device will
+			 * cease to function. Is fatal error appropriate here?
+			 */
+			return;
+		}
+
+		sys_write32((uint32_t)buf->data,
+			    (mem_addr_t)&base->out_ep[ep_idx].doepdma);
+
+		sys_cache_data_invd_range(buf->data, xfersize);
```

Comment:
It not the same range invalidated twice, here and in smaller blocks at L:947?

Reply:
The problem with dirty cache lines is that they can be evicted at any time regardless of software actions. The invalidation here is to make sure there are no dirty cache lines spanning over the buffer. This invalidation must be done before the DMA transfer is started.

The other invalidation done in `dwc2_handle_out_xfercompl()` is to ensure that software did not cache any data  during the transfer (and if it did, then it is discarded).

Reply:
> The other invalidation done in `dwc2_handle_out_xfercompl()` is to ensure that software did not cache (read) any data during the transfer (and if it did, then it is discarded).

But no one should access the buffer as long it is queued for transfer.
To invalidate a range again seems to be superfluous for me, if someone writes in between has a problem anyway, right?

Reply:
If someone writes, then yes it is a problem - but this problem is outside of a scope here. The invalidate in `dwc2_handle_out_xfercompl()` is for discarding potential cache entries that did occur due to read.

Reply:
> The invalidate in `dwc2_handle_out_xfercompl()` is for discarding potential cache entries that did occur due to read.

Cache invalidate in dwc2_handle_out_xfercompl() seems  to be reasonable, but what could happen if we start DMA but do not invalidate?

Reply:
"write to it" is a bit confusing, could also be a read access, but okay. In the future, would it make sense for the allocator to invalidate the hole block?

Reply:
Read access wouldn't pose any problem, because the eventual cache eviction of "clean" line is not going to result in write.

I think the allocator should be just dealing with alignment and padding and should leave the cache invalidation/flushing to the actual user.